core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java

```diff
@@ -5451,7 +5461,11 @@ ImmutableList<RelNode> retrieveCursors() {
       case CURSOR:
       case IN:
       case NOT_IN:
-        subQuery = requireNonNull(getSubQuery(expr, null));
+        subQuery = getSubQuery(expr, null);
+        if (subQuery == null && (kind == SqlKind.SOME || kind == SqlKind.ALL)) {
+          break;
+        }
+        assert subQuery != null;
```

Comment:
This assert can probably be removed as requireNonNull is called right after.

Reply:
Probably it can be removed. However, I feel that it's important to keep it because requireNonNull is applied to `subQuery.expr`. If a NullPointerException is thrown, it could be difficult to identify whether the root cause lies with `subQuery` or `subQuery.expr`.

Reply:
Agree with @bchapuis, it is meaningless to call 'assert' here.

Reply:
@zoudan @bchapuis  I tend to disagree with you guys. How to distinguish NPE in these pieces of code without debugging ?
```
subQuery = null;
// .... Some code
requireNonNull(subQuery.expr);
```

```
subQuery = new SubQuery(node, logic, clause);
subQuery.expr = null;
// .... Some code
requireNonNull(subQuery.expr);
```

Both of these snippets result in an NPE with an identical stacktrace. The only way to understand what the root cause of the NPE (e.g `subQuery` or `subQuery.expr`) is debugging this code. However, getting additional information (e.g., queries) for debugging can be hard in a production environment due to security concerns. So, It's really important to have as much information as possible to understand and fix the problem.